log.go

```diff
@@ -80,34 +98,197 @@ type eventLogger struct {
+		return
+	}
+	if err == nil {
+		return
+	}
+
+	otExt.Error.Set(span, true)
+	span.LogKV("error", err.Error())
+}
+
+// Finish completes the span associated with `ctx` by
+// setting the end timestamp and finalizing the Span state.
+//
+// Finish() must be the last call made to any span instance, and to do
+// otherwise leads to undefined behavior.
+// Finish will do its best to notify (log) when used in correctly
+//		.e.g called twice, or called on a spanless `ctx`
+func (el *eventLogger) Finish(ctx context.Context) {
+	span := opentrace.SpanFromContext(ctx)
+	if span == nil {
+		_, file, line, _ := runtime.Caller(1)
+		log.Errorf("Finish with no Span in context called on %s:%d", path.Base(file), line)
+		return
+	}
+
+	//programmer error if this fails to cast - no error check
+	val := ctx.Value(activeSpanKey)
+	as := val.(*activeSpan)
+	if as.isFinished {
+		_, file, line, _ := runtime.Caller(1)
+		log.Errorf("The span has already been finished in %s:%d", path.Base(file), line)
+	} else {
+		span.Finish()
+		as.isFinished = true
+	}
+}
+
+// FinishWithErr completes the span associated with `ctx` by
+// setting the end timestamp and finalizing the Span state, in addition to
+// calling the `SetErr` method first
+//
+// FinishWithErr() must be the last call made to any span instance, and to do
+// otherwise leads to undefined behavior.
+// FinishWithErr will do its best to notify (log) when used in correctly
+//		.e.g called twice, or called on a spanless `ctx`
+func (el *eventLogger) FinishWithErr(ctx context.Context, err error) {
+	el.SetErr(ctx, err)
+	el.Finish(ctx)
+}
+
+func deserializeContext(bCtx []byte) (opentrace.SpanContext, error) {
+	gTracer := opentrace.GlobalTracer()
+	carrier := bytes.NewReader(bCtx)
+	spanContext, err := gTracer.Extract(opentrace.Binary, carrier)
+	if err != nil {
+		log.Warning("Failed to deserialize context %s", err)
+		return nil, err
+	}
+	return spanContext, nil
+}
+
+// DEPRECATED use `Start`
 func (el *eventLogger) EventBegin(ctx context.Context, event string, metadata ...Loggable) *EventInProgress {
-	start := time.Now()
-	el.Event(ctx, fmt.Sprintf("%sBegin", event), metadata...)
+	ctx = el.Start(ctx, event)
 
-	span, ctx := opentrace.StartSpanFromContext(ctx, event)
+	for _, m := range metadata {
+		for l, v := range m.Loggable() {
+			el.LogKV(ctx, l, v)
+		}
+	}
 
 	eip := &EventInProgress{}
 	eip.doneFunc = func(additional []Loggable) {
-		metadata = append(metadata, additional...)                      // anything added during the operation
-		metadata = append(metadata, LoggableMap(map[string]interface{}{ // finally, duration of event
-			"duration": time.Now().Sub(start),
-		}))
-
-		el.Event(ctx, event, metadata...)
-		if traceingDisabled() {
-			return
-		}
-		otExt.Component.Set(span, el.system)
+		// anything added during the operation
+		// e.g. deprecated methods event.Append(...) or event.SetError(...)
+		metadata = append(metadata, additional...)
```

Comment:
This looks to add the same loggables twice from the original arguments, I think this probably was suppose to be:

`metadata = append(eip.loggables, additional...)`

Another possible way to handle the integration with the `EventInProgress` type would be to store the `eventLogger` / `ctx` and directly work on those, following your `DEPRECATED` comments.

Reply:
dohh, good catch